A few tractometry functions

arokem opened the pull request:
For tractometry, we want to be able to extract the profiles along the length of a bundle 

This PR introduces a couple of useful functions for that: 

- `bundle_profile` takes a bundle and returns the values along the length of the bundle, resampled to 100 (or some other number of) nodes. 
- `gaussian_weights` allows you to weight the streamlines within a bundle, depending on how far they are from the core of the bundle.
- Finally, `orient_by_streamline` will make sure that a bundle is all oriented in the same direction (using the other functions without doing this will result in erroneous results!), based on the trajectory of a "standard" streamline. 

Thanks to @jyeatman for guidance in implementing these in the prototype implementation we are using in pyAFQ!

skoudoro commented on dipy/tracking/streamline.py:
```diff
@@ -649,7 +650,82 @@ def orient_by_rois(streamlines, roi1, roi2, in_place=False,
+    affine : ndarray
+        Affine transformation from voxels to streamlines. Default: identity.
+
+    Returns
```
Can you complete this returns

arokem replied:
Resolved in d18c9e84

skoudoro commented on dipy/tracking/streamline.py:
```diff
@@ -649,7 +650,82 @@ def orient_by_rois(streamlines, roi1, roi2, in_place=False,
+    """
+    # Start by resampling into an array, so that distance calculation is easy:
+    fgarray = np.array(set_number_of_points(streamlines,  n_points))
+    std_array = np.array(set_number_of_points([standard],  n_points))
```
I recommend removing this `np.array`

arokem commented:
I am trying to create an example, which would follow the bundle extraction example, using these bundles to do tractometry. For that, I will also need parameter maps for this subject. 

@Garyfallidis: Which of the HCP subjects was used to create the `fetch_target_tractogram_hcp` tractogram?

arokem commented:
Sorry - @BramshQamar : I see that you made that tutorial. Do you know which subject `fetch_target_tractogram_hcp` belongs to?

BramshQamar commented:
> Sorry - @BramshQamar : I see that you made that tutorial. Do you know which subject `fetch_target_tractogram_hcp` belongs to?

Hello @arokem, we used subject 100206 from HCP data to create target tractogram for bundle extraction example.

pep8speaks commented:
Hello @arokem, Thank you for updating !





 - In the file [`dipy/data/fetcher.py`](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/data/fetcher.py), following are the PEP8 issues :

> [Line 451:81](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/data/fetcher.py#L451): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)



 - In the file [`dipy/tracking/streamline.py`](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/tracking/streamline.py), following are the PEP8 issues :

> [Line 724:5](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/tracking/streamline.py#L724): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 991:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/tracking/streamline.py#L991): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

 - In the file [`dipy/tracking/tests/test_streamline.py`](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/tracking/tests/test_streamline.py), following are the PEP8 issues :

> [Line 1002:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/tracking/tests/test_streamline.py#L1002): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 1310:5](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/dipy/tracking/tests/test_streamline.py#L1310): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

 - In the file [`doc/examples/bundle_extraction.py`](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py), following are the PEP8 issues :

> [Line 116:44](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L116): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 117:44](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L117): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 118:44](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L118): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 119:44](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L119): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 120:44](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L120): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 121:44](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L121): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 132:42](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L132): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 132:45](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L132): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 133:47](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L133): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 133:50](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L133): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 157:10](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L157): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '('
> [Line 162:45](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L162): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 163:45](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L163): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 164:45](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L164): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 165:45](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L165): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 166:45](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L166): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 167:45](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L167): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 178:43](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L178): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 178:46](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L178): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 179:48](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L179): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 179:51](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L179): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 215:4](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_extraction.py#L215): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

 - In the file [`doc/examples/bundle_profiles.py`](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py), following are the PEP8 issues :

> [Line 44:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L44): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 64:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L64): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 65:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L65): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 72:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L72): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 73:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L73): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 103:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L103): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 115:81](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L115): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 122:1](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L122): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 139:74](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L139): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 139:74](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L139): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 142:72](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L142): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 142:72](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L142): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 174:4](https://github.com/nipy/dipy/blob/dd5ae08bea8eb8acb515135570202393dc2604af/doc/examples/bundle_profiles.py#L174): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +878,114 @@ def values_from_volume(data, streamlines, affine=None):
+        n_points = bundle.shape[1]
+    else:
+        # It's something else, assume that it needs to be resampled:
+        bundle = np.array(set_number_of_points(bundle, n_points))
```
I believe this makes some of the operations below impossible. See L925, for example.

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +881,119 @@ def values_from_volume(data, streamlines, affine=None):
+        # This should come back as a 3D covariance matrix with the spatial
+        # variance covariance of this node across the different streamlines
+        # This is a 3-by-3 array:
+        node_coords = bundle[:, node]
```
Sorry - it moved now. This is the line I referred to in my previous comment (https://github.com/nipy/dipy/pull/1695/files#r245489936). I don't think that I would be able to do this with something that isn't an array. But I'd be happy to learn that I am wrong.

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +878,114 @@ def values_from_volume(data, streamlines, affine=None):
+    # should be weighted)
+    w = 1 / w
+    # Normalize before returning, so that the weights in each node sum to 1:
+    return w / np.sum(w, 0)
```
Do you have something particular in mind? 

I think that Mahalnobis is strictly non-negative, and so I think that all the mahalnobis distances on a node would only be 0 if all the streamlines in the bundle are identical in that node. 

I am adding a test for that situation, and that throws a linalg error earlier in the function earlier on, because the covariance matrix is all zeros. I think that I should catch it there instead.

The question is what to do in this case. I think that the right thing would be to give each streamline and identical weight that is 1/n_streamlines, which would mean that they would all contribute equally and sum to the value in that voxel. So that's what I will implement for now. Let me know if you think that's wrong.

arokem replied:
Now implemented in b7f61a5e

skoudoro replied:
> which would mean that they would all contribute equally and sum to the value in that voxel

Nice! I did not think about all of that but it all makes sense to me and I agree.  Thank you for this update!

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +878,114 @@ def values_from_volume(data, streamlines, affine=None):
+    Parameters
+    ----------
+    data : 3D volume
+    bundle : StreamLines class instance, list of arrays, or array
```
OK. Changed here and in the docstring above.

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +878,114 @@ def values_from_volume(data, streamlines, affine=None):
+        Weight each streamline (1D) or each node (2D) when calculating the
+        tract-profiles. Must sum to 1 across streamlines (in each node if
+        relevant).
+
```
Added in d18c9e84

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +878,114 @@ def values_from_volume(data, streamlines, affine=None):
+        # list => np.array
+        fgarray = np.array(set_number_of_points(bundle, n_points))
+
+    values = values_from_volume(data, fgarray, affine=affine)
```
The purpose of the resampling is to use all of the streamlines in the bundle to sample from the volume. But to do so, we need to make sure that all the streamlines have the same number of nodes, so that they roughly (!) correspond to each other. 

The approach is described in some detail in this paper: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0049790, which I will add as a reference.

skoudoro replied:
Gotcha! it is clear to me for `fgarray` 

I still think that you do not need to cast `set_number_of_points(bundle, n_points)` with  `np.array`.

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +878,114 @@ def values_from_volume(data, streamlines, affine=None):
+    values = values_from_volume(data, fgarray, affine=affine)
+
+    if weights is None:
+        weights = np.ones(values.shape) / values.shape[0]
```
I believe that this would only happen if there are no streamlines in the bundle. Should we catch that and raise an error at the top?

skoudoro replied:
yes, it would be great, even if I think it will almost never happen...

skoudoro replied:
No, looks good! Personally, I prefer to get an error message than a silent empty result.

arokem commented:
I added an example usage in f727bde2. I had to introduce some changes in the bundle extraction example, because I am reusing the bundles generated in that example.

skoudoro commented on doc/examples/bundle_extraction.py:
```diff
@@ -92,23 +102,23 @@
+                                           reduction_distance='mam',
+                                           slr=True,
+                                           slr_metric='asymmetric',
+                                           pruning_distance='mam')
```
pep8: alignment

skoudoro commented on doc/examples/bundle_extraction.py:
```diff
@@ -136,15 +146,25 @@
+                                            reduction_distance='mam',
+                                            slr=True,
+                                            slr_metric='asymmetric',
+                                            pruning_distance='mam')
```
pep8: alignment

skoudoro commented on dipy/tracking/streamline.py:
```diff
@@ -802,5 +880,133 @@ def values_from_volume(data, streamlines, affine=None):
+        # This should come back as a 3D covariance matrix with the spatial
+        # variance covariance of this node across the different streamlines
+        # This is a 3-by-3 array:
+        node_coords = bundle.data[node::n_points]
```
`[node::n_points]` Nice! 👍

skoudoro commented:
After a rebase, this PR is ready to go, it will be good if someone could have a look on this one

arokem commented:
Now rebased!

On Mon, Jan 14, 2019 at 8:55 AM Serge Koudoro <notifications@github.com>
wrote:

> After a rebase, this PR is ready to go, it will be good if someone could
> have a look on this one
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1695#issuecomment-454078185>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNmqyyHZzBPIOvrp2Hx_zC_VtgfVhks5vDLabgaJpZM4ZXTgw>
> .
>

arokem commented:
Any more thoughts here?

arokem commented:
Oh - sorry about that - should be fixed now.

On Fri, Feb 1, 2019 at 7:29 AM Serge Koudoro <notifications@github.com>
wrote:

> *@skoudoro* commented on this pull request.
>
> I will merge this on Wednesday if there is no other comment.
>
> Can you fix the small rebase residue below @arokem
> <https://github.com/arokem>?
> ------------------------------
>
> In doc/examples_index.rst
> <https://github.com/nipy/dipy/pull/1695#discussion_r253088796>:
>
> > @@ -167,6 +167,10 @@ Streamline analysis and connectivity
>  - :ref:`example_streamline_length`
>  - :ref:`example_cluster_confidence`
>  - :ref:`example_path_length_map`
> +<<<<<<< HEAD
>
> rebase residue
> ------------------------------
>
> In doc/examples_index.rst
> <https://github.com/nipy/dipy/pull/1695#discussion_r253088893>:
>
> > @@ -167,6 +167,10 @@ Streamline analysis and connectivity
>  - :ref:`example_streamline_length`
>  - :ref:`example_cluster_confidence`
>  - :ref:`example_path_length_map`
> +<<<<<<< HEAD
> +=======
> +- :ref:`example_bundle_profiles`
> +>>>>>>> Adds an example of bundle profile extraction (tractometry)
>
> rebase residue
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1695#pullrequestreview-199125047>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNutWecKLPDMpOyUHAeeJc6vv3pbgks5vJF1mgaJpZM4ZXTgw>
> .
>

arokem commented:
Does anyone else want to take a look? Any reason not to merge this?

skoudoro commented:
ping @Garyfallidis

arokem commented:
Looks like a new release of cvxpy is breaking MAPMRI.

skoudoro commented:
I just did a quick check on `cvxpy` changelog and there are very few changes. Mapmri tests seems really sensitive... I will have a deeper look later

skoudoro commented:
Hi @arokem, I just created this issue https://github.com/cvxgrp/cvxpy/issues/672 for our `cvxpy` problem. Concerning Travis,  I do not know what should be our strategy to make Travis happy until their fix

arokem commented:
Pin the cvxpy version to the previous release for now?

arokem commented:
Implemented in https://github.com/nipy/dipy/pull/1695/commits/cbf9275bf68bf8879416b7bbfa218eb48f2c9212, I think.

skoudoro commented:
Great! Thanks! we have to put a reminder somewhere for coming back to normal later

arokem commented:
OK. Now rebased to incorporate #1740

arokem commented:
@Garyfallidis : do you intend to give this a look through?

arokem commented on dipy/tracking/streamline.py:
```diff
@@ -801,5 +879,133 @@ def values_from_volume(data, streamlines, affine=None):
+    return w / np.sum(w, 0)
+
+
+def bundle_profile(data, bundle, affine=None, n_points=100,
```
From @Garyfallidis : Consider renaming to something that doesn't have `profile` in it (to avoid confusing with profilometry).

arokem replied:
Upon re-reading the [profilometry paper](https://www.ncbi.nlm.nih.gov/pubmed/26667008), I am not convinced that it's all that different. The first step in that analysis is also the extraction of bundle profiles, just like what we're doing here. Then, they do a whole lot of multivariate analysis on top of it, but you could do that with the profiles we generate here as well. 

WDYT, @Garyfallidis?

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,175 @@
+"""
+========================================================
+Tractometry: extracting profiles from segmeneted bundles
```
typo

Garyfallidis commented on dipy/tracking/tests/test_streamline.py:
```diff
@@ -1184,5 +1261,118 @@ def test_cluster_confidence():
+
+def test_bundle_profile():
+    data = np.ones((10, 10, 10))
+    bundle = Streamlines([np.array([[0, 0., 0],
```
Try this
bundle = Streamlines()
bundle.extend(A)
bundle.extend(B)

avoid using lists of numpy arrays.

Garyfallidis commented on dipy/tracking/tests/test_streamline.py:
```diff
@@ -1184,5 +1261,118 @@ def test_cluster_confidence():
+    affine = np.eye(4)
+    affine[:, 3] = [-1, 100, -20, 1]
+    # Transform the streamlines:
+    x_bundle = Streamlines([sl + affine[:3, 3] for sl in bundle])
```
Same here. Use extend. Or even better transform the data directly using
x_bundle._data

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,175 @@
+    Brian A. Wandell, and Heidi M. Feldman. 2012. "Tract Profiles of White
+    Matter Properties: Automating Fiber-Tract Quantification" PloS One 7
+    (11): e49790.
+
```
Add missing references. For example reference for QB.

Garyfallidis commented:
Ping @arokem :)

arokem commented:
Sorry. I am in grant writing prison until Monday. Will get back to this then

On Fri, Feb 22, 2019 at 8:07 AM Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Ping @arokem <https://github.com/arokem> :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1695#issuecomment-466449899>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNlz9tZZdFdu14u5IYUT2DY5uIyL1ks5vQBXFgaJpZM4ZXTgw>
> .
>

Garyfallidis commented:
Sure okay. Good luck!

arokem commented:
@Garyfallidis : I think that I addressed all your comments in https://github.com/nipy/dipy/pull/1695/commits/6f65734e13e9b55d98e8cdc0b73b2ebc5e816afe

Let me know what you think!

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+"""
+========================================================
+Tractometry: extracting profiles from segmented bundles
```
Suggesting alternative title (this is too generic).

Extracting AFQ-style profiles from segmented bundles.

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+
+In this example, we will we will extract the values of a statistic from a
+volume, using the coordinates along the length of a bundle. This is called
+tractometry.
```
This is a type of tractometry. 
Tractometry does not always measure along the the length of bundles.

arokem replied:
What else does tractometry do?

I think the term comes from this ISMRM presentation: https://cds.ismrm.org/protected/11MProceedings/files/678.pdf 

That's also the way it's used in this paper: https://www.frontiersin.org/articles/10.3389/fninf.2017.00042/full

arokem replied:
I've changed this, but I still think that tractometry is the right term here. Let me know what you think.

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+
+"""
+
+In the next step, we need to make sure that all the streamlines in each bundles
```
each bundle

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+
+Since we are going to include all of the streamlines in the single cluster
+from the streamlines, we set the threshold to `np.inf`. We pull out the centroid
+as the standard
```
full stop missing

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+affine used is the one from the transform between the atlas and individual
+tractogram. This is so that the orienting is done relative to the space of the
+individual, and not relative to the atlas space.
+
```
remove space

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+
+
+"""
+
```
remove space

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+And then use the weights to calculate the bundle profiles for each bundle
+"""
+
+profile_cst_l = dts.bundle_profile(fa, oriented_cst_l, affine=img.affine)#,
```
rename bundle_profile to afq_profile

Garyfallidis replied:
Please move afq_profile to dipy.segment.bundle module.

Garyfallidis replied:
The afq_profile could have done both the orientation and weights calculation. Not sure why you need independent functions for these.

Garyfallidis commented on dipy/align/streamlinear.py:
```diff
@@ -290,14 +290,14 @@ def __init__(self, metric=None, x0="rigid", method='L-BFGS-B',
         .. [Garyfallidis15] Garyfallidis et al. "Robust and efficient linear
-            registration of white-matter fascicles in the space of
-            streamlines", NeuroImage, 117, 124--140, 2015
+                            registration of white-matter fascicles in the space
```
Not sure about the tab shift here. Is it really needed?

arokem replied:
I think you're right!

Garyfallidis commented on dipy/tracking/streamline.py:
```diff
@@ -801,5 +879,133 @@ def values_from_volume(data, streamlines, affine=None):
+    return w / np.sum(w, 0)
+
+
+def bundle_profile(data, bundle, affine=None, n_points=100,
```
See my other comment below. Suggest renaming to afq_profile and move to dipy.segment.bundles

Garyfallidis replied:
This function could also automatically calculate orient and weights. You could add flags inside this function for that. if orient==True etc.

arokem replied:
It makes sense to orient separately, because you might want to use the same bundle to extract different metrics (FA, MD and so forth) and you don't want to have to orient them each time. Also, if you want to save the oriented tracks, you'll need to have them outside of the function. 

That said, you're right that these can be options. Adding as key-word arguments!

Garyfallidis commented on dipy/tracking/streamline.py:
```diff
@@ -801,5 +879,133 @@ def values_from_volume(data, streamlines, affine=None):
+        The collection of streamlines (possibly already resampled into an array
+         for each to have the same length) with which we are resampling. See
+         Note below about orienting the streamlines.
+    weights : 1D
```
1D what?

Garyfallidis commented:
Okay see my last comments. And please correct all last Pep8 errors. You have quite a few. And thanks for all the cool work.

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+profile_cst_l = dts.bundle_profile(fa, oriented_cst_l, affine=img.affine)#,
+#                                   weights=w_cst_l)
+
+profile_af_l = dts.bundle_profile(fa, oriented_af_l, affine=img.affine)#,
```
Comments need deleting

Garyfallidis commented on doc/examples/bundle_profiles.py:
```diff
@@ -0,0 +1,183 @@
+ax1.set_ylabel("Fractional anisotropy")
+ax1.set_xlabel("Node")
+ax2.plot(profile_af_l)
+ax2.set_xlabel("Node")
```
Both labels say Node!

Garyfallidis replied:
You will probably want to plot the std too as in the paper?

arokem replied:
Standard deviation is usually calculated across subjects, so it's zero here.

arokem commented:
Thanks for the review. I think that I have addressed all the comments. 

I renamed the function `afq_tract_profiles`, which is how they are called in the original Yeatman paper. 

Let me know what you think!

arokem commented:
Now also rebased.

Garyfallidis commented:
@arokem ready for final check?

arokem commented:
I believe so!

Garyfallidis commented on dipy/segment/bundles.py:
```diff
@@ -632,3 +635,153 @@ def _prune_what_not_in_model(self, model_centroids,
+    return w / np.sum(w, 0)
+
+
+def afq_tract_profile(data, bundle, affine=None, n_points=100,
```
You are already in bundles.py no need for tract. I would just say afq_profile.

Garyfallidis commented:
I added a last comment. Please correct and ping me to merge! :)

Garyfallidis commented:
In general I use bundles to mean tracts, fascicles, muscle fibers etc. The tools are quite generic so I think bundles is good name to use. I don't think we need to specify tracts anymore. However, in the tutorial it is good to have some anatomical naming standards. Which you do :+1:

arokem commented:
Yeah - good call. Renamed!
